### PR TITLE
Minor improvements

### PR DESCRIPTION
**Pull request by tuexss**
I tried to clean it up a bit, by having better formatting and less variable-spaghetti. I got rid off the Class-Inspection and replaced with respond_to?

I'd like to discuss the necessity of the sorted parameter, because the algorithm assumes a sorting, so it would make sense to always, sort. And if it's already sorted no harm will be done either.

### CODE REVIEW

**Comment by tuexss** [comment]
btw: have you considered using http://ruby-doc.org/core-2.2.3/Array.html#method-i-bsearch and just passing the range-check as a block? This _is_ an array after all, as of now. Or maybe we should try implementing a real segment tree again, find out what's the issue with the memory.

File: lib/segment_tree.rb [hunk 1]
```diff
@@ -84,17 +78,9 @@ def inspect
...
-    left      = idx > 0 && @segments[idx - 1]
-    right     = idx < @segments.size - 1 && @segments[idx + 1]
-
-    case
-      when left && low.range.begin <= x && x <= left.range.end then -1
-      when segment.range.begin <=x && x <= segment.range.end then 0
-      when right && right.range.begin <=x && x <= high.range.end then 1
-      else nil
-    end
+  def matches?(x, low_idx, idx, high_idx) #:nodoc:
```
Comment (take-five) [inline lib/segment_tree.rb]: I don't really like how code looks now. That variables you've removed were mostly for self-documenting. Now code is harder to understand

File: lib/segment_tree.rb [hunk 2]
```diff
@@ -47,13 +47,8 @@ def <=>(other)
...
   def initialize(data, sorted = false)
     # build elementary segments
-    @segments = case data
-      when Hash, Array, Enumerable then
-        data.collect { |range, value| Segment.new(range, value) }
-      else raise ArgumentError, '2-dim Array or Hash expected'
-    end
-
-    @segments.sort! unless sorted
+    raise ArgumentError, '2-dim Array or Hash expected' unless data.respond_to?('collect')
```
Comment (take-five) [inline lib/segment_tree.rb]: This change makes sense! :+1:

File: lib/segment_tree.rb (hunk 1 above)
Comment (tuexss) [inline lib/segment_tree.rb]: Yeah, not too happy with it either. but as you can see on the link above (there's a small link 'click to toggle source' in the top right corner when you hover with the mouse), this whole algorithm is a duplication of what's already in bsearch. so maybe use that one specialize on the specialities for ranges, like inclusive vs. exclusive boundaries. wdyt?

Comment (take-five) [inline lib/segment_tree.rb]: Yes, `Array#bsearch` probably can do the job. Do you want to implement it yourself?

Comment (tuexss) [inline lib/segment_tree.rb]: I'll try it this evening.

**Pull request closed without merging**